lms/services/blackboard_api/_basic.py

```diff
@@ -64,22 +64,39 @@ def __init__(
 
     def get_token(self, authorization_code):
         self._oauth_http_service.access_token_request(
-            token_url=self._api_url("oauth2/token"),
+            token_url=self.token_url,
             redirect_uri=self.redirect_uri,
             auth=(self.client_id, self.client_secret),
             authorization_code=authorization_code,
         )
 
     def request(self, method, path):
-        try:
-            return self._oauth_http_service.request(method, self._api_url(path))
-        except HTTPError as err:
-            error_dict = BlackboardErrorResponseSchema(err.response).parse()
+        url = self._api_url(path)
+
+        def send():
+            try:
+                return self._oauth_http_service.request(method, url)
+            except HTTPError as err:
+                error_dict = BlackboardErrorResponseSchema(err.response).parse()
 
-            if error_dict.get("message") == "Bearer token is invalid":
-                raise OAuth2TokenError() from err
+                if error_dict.get("message") == "Bearer token is invalid":
+                    raise OAuth2TokenError() from err
 
-            raise
+                raise
+
+        try:
+            return send()
+        except (OAuth2TokenError, HTTPError):
+            self._oauth_http_service.refresh_token_request(
+                self.token_url,
+                self.redirect_uri,
+                auth=(self.client_id, self.client_secret),
+            )
+            return send()
```

Comment:
I originally wanted to add generic "refresh and retry" support to `OAuthHTTPService` but I ended up implementing it here in the Blackboard API client instead. Do we know that we'll always want refresh-and-retry, for every OAuth API? And I don't think a `OAuthHTTPService.request(..., retry=True)` boolean to turn if off would suffice either: the decision about whether or not to refresh-and-retry may depend on API-specific details of the error response, we may want to refresh-and-retry for some errors but not others. You can see in the code above that some Blackboard-specific error parsing is necessarily involved.

It may be worth taking another look at Blackboard's `"Bearer token is invalid"` error response and seeing whether there's any standard OAuth 2 fields in it that could be used to identify it as an access token error. But I don't think there are.

Reply:
Hmm, yeah, I think it might be possible to split out generic refresh-and-retry logic into `OAuthHTTPService.request()` and leave the Blackboard-specific error handling here in `BasicClient.request()`.

If we do that then the `OAuthHTTPService.request()` unit tests will have to use `patch.object(...)` to mock `OAuthHTTPService.refresh_token_request()` but I think that's fine.